### DISCUSSION TITLE
QRCODE reading and event triggering

### DISCUSSION THREAD

**Discussion by tyhudson** [AxisCommunications/discussions#222] [General] [answered]
I have an AXIS P1468-LE Bullet Camera with a barcode and QRcode reading license. 

I have also set up a rule to detect barcodes. 

![image](https://user-images.githubusercontent.com/118188666/231861608-d35758f8-3e9b-4838-83c8-a6aa7295d93a.png)

This is working as it should to hit a custom HTTP endpoint.  The issue I am having is I can not send the barcode or QR code data to the endpoint. 

![image](https://user-images.githubusercontent.com/118188666/231863587-02345b8a-1613-4c26-a951-56e8bb517df6.png)

I attempted to send the barcode or qrcode just like the dynamic overlay with the #D but it does not work.

**Comment by daniel-falk** [comment, marked as answer]
Hi Ty,

I do not think it is possible to send the event data in the POST actions, I am not 100% sure so it would be great if someone from Axis could confirm or correct @shadysp. There are a few ways I can think of to get the data you want:


**Use the RTSP event stream**
--------------------------------------
All event info is streamable over RTSP, e.g. from the bar-code reader you will get:
```
<?xml version="1.0" encoding="UTF-8"?>
<tt:MetadataStream xmlns:tt="http://www.onvif.org/ver10/schema">
<tt:Event><wsnt:NotificationMessage xmlns:tns1="http://www.onvif.org/ver10/topics" xmlns:tnsaxis="http://www.axis.com/2009/event/topics" xmlns:wsnt="http://docs.oasis-open.org/wsn/b-2" xmlns:wsa5="http://www.w3.org/2005/08/addressing"><wsnt:Topic Dialect="http://docs.oasis-open.org/wsn/t-1/TopicExpression/Simple">tnsaxis:CameraApplicationPlatform/BarcodeReader/BarcodeFound</wsnt:Topic><wsnt:ProducerReference><wsa5:Address>uri://9a7f75df-99ca-409b-a88d-a0ac664be538/ProducerReference</wsa5:Address></wsnt:ProducerReference><wsnt:Message><tt:Message UtcTime="2023-04-14T09:29:39.385120Z"><tt:Source></tt:Source><tt:Data><tt:SimpleItem Name="type" Value="QRCode"/><tt:SimpleItem Name="time" Value="2023-04-14 11:29:39.027"/><tt:SimpleItem Name="data" Value="https%3A%2F%2Fwww.investopedia.com%2Fterms%2Fq%2Fquick-response-qr-code.asp"/></tt:Data></tt:Message></wsnt:Message></wsnt:NotificationMessage></tt:Event></tt:MetadataStream>
```
Where in particular you are interested in the data field:
```
<tt:SimpleItem Name="data" Value="..."/></tt:Data>
```
The problem with this is that you need to have network access to the camera.


**Use MQTT**
----------------
You can push the events using the MQTT client in the camera. This is an example of the QR event being published to a MQTT broker:
```
axis/B8A44F3B41FC/event/CameraApplicationPlatform/BarcodeReader/BarcodeFound {"topic":"axis:CameraApplicationPlatform/BarcodeReader/BarcodeFound","timestamp":1681466022028,"serial":"B8A44F3B41FC","message":{"source":{},"key":{},"data":{"type":"QRCode","data":"https%3A%2F%2Fwww.investopedia.com%2Fterms%2Fq%2Fquick-response-qr-code.asp","time":"2023-04-14 11:53:41.680"}}}
```


**Create a custom ACAP for event push**
---------------------------------------------------
It would be easy to create a stand-alone ACAP that subscribes to the events and pushed them to an HTTP/HTTPS server. We have templates for event consumer ACAPs and precompiled libraries for cURL and OpenSSL, you can find more info at https://fixedit.ai. In your situation this might be the best option. The downside is that you need to install two different ACAPs but on the other hand you will get exactly the kind of integration you want to have. We can assist you or provide the implementation for you. I'll send you an email with more info if you want to go forward with this.

For reference, the code to subscribe to the QR code event would be like this:
```c
AXEventKeyValueSet *key_value_set = ax_event_key_value_set_new();
if (!ax_event_key_value_set_add_key_values(
          key_value_set, NULL,
          "topic0", "tnsaxis", "CameraApplicationPlatform",  AX_VALUE_TYPE_STRING,
          "topic1", "tnsaxis", "BarcodeReader",  AX_VALUE_TYPE_STRING,
          "topic2", "tnsaxis", "BarcodeFound",   AX_VALUE_TYPE_STRING,
          "data", NULL, NULL, AX_VALUE_TYPE_STRING,
          error)) {
    ....
}

...

if (!ax_event_key_value_set_get_string(key_value_set, "data",
          NULL /* namespace */, &qr_code, &error)) {
    ...
}

fapp_logger_log(LOG_INFO, "Event received: qr_code=%s", qr_code);
g_free(qr_code);
```

**Reply by tyhudson** [reply]
WOW, great response.  Thank you so much let me take a look and see what I can do. 

Once again, thank you.

**Reply by tyhudson** [reply]
What I ended up doing is setting the camera to connect to the MQTT broker.
